lib/Dialect/FIRRTL/FIRRTLFolds.cpp

```diff
@@ -1993,7 +2036,188 @@ LogicalResult MemOp::canonicalize(MemOp op, PatternRewriter &rewriter) {
     }
   }
   rewriter.eraseOp(op);
-  return success();
+  return true;
+}
+
+// If memory has no write ports, eliminate it.
+bool foldReadOrWriteOnlyMemory(PatternRewriter &rewriter, MemOp op) {
+  bool isRead = false, isWritten = false;
+  for (unsigned i = 0; i < op.getNumResults(); ++i) {
+    switch (op.getPortKind(i)) {
+    case MemOp::PortKind::Read:
+      isRead = true;
+      if (isWritten)
+        return false;
+      continue;
+    case MemOp::PortKind::Write:
+      isWritten = true;
+      if (isRead)
+        return false;
+      continue;
```

Comment:
What happens if the memory is written but not read ?

Reply:
Then `isWritten = true` and `isRead = false`, the memory is write-only and can be erased.